File: Strokes/Assets/StreamingAssets/Stroke.cpp
#include "Stroke.h"
#include "cinder/Rand.h"

namespace rph {
    
Stroke::Stroke(const Options &options)
{
    mOpts = options;
}
    
Stroke::Stroke(const std::vector<vec2> &points, const Options &options)
{
    mOpts = options;
    
    for (auto &point : points) {
        if (mInputPoints.empty()) {
            start(point);
        }
        else {
            move(point);
        }
    }
    end();
}

void Stroke::start(const vec2 &pos)
{
    mDrawing = true;
    mStartTime = app::getElapsedSeconds();
    mBounds = Rectf(pos, pos);
    move(pos);
}

void Stroke::move(const vec2 &pos)
{
    if (!mDrawing) return;
    
    if (mInputPoints.size() <= 2) {
        mInputPoints.push_back(pos);
    }
    else {
        vec2 newPos = lerp(mInputPoints.back(), pos, 0.35f);
        
        // calculate thickness
        float dist = glm::distance(newPos, mInputPoints.back());
        
        if (dist < 1.0) return;
        
//        float newThick = clamp(mOpts.baseThickness + dist * mOpts.speedToThickness, mOpts.minThickness, mOpts.maxThickness);
        float newThick = dist; // moved to updategeometry
        float prevThick = mThickness.empty() ? newThick : mThickness.back();
        float currThick = lerp(prevThick, newThick, 0.2);
        
        // get last 3 input points
        const vec2 &prev2 = mInputPoints.at(mInputPoints.size() - 2);
        const vec2 &prev1 = mInputPoints.at(mInputPoints.size() - 1);
        const vec2 &cur   = newPos;
    
        // create cubic bezier segment with inputs as control points
        Path2d path;
        path.moveTo((prev2 + prev1) / 2.0f);
        path.quadTo(prev1, (prev1 + cur) / 2.0f);
       
        // divide segment adaptatively depending on its length
        // save vertices and thickness
        float pathLength = path.calcLength();
        int divisions = ceil(pathLength / mMinSegmentLength);
        
        for (int i = 1; i <= divisions; i++) {
            float t = float(i) / float(divisions);
            float thick = lerp(prevThick, currThick, t);
            vec2 point = path.getPosition(t);
            vec2 norm  = glm::normalize(newPos - (mSampledPoints.size() ? mSampledPoints.back() : mInputPoints.back()));
            vec2 perp  = vec2(-norm.y, norm.x);
            
            mSampledPoints.push_back(point);
            mThickness.push_back(thick);
            mOffsets.push_back(perp);
            
            //CI_LOG_D("interp: " << count << " t: " << t);
            //CI_LOG_D("interp: " << count << " x: " << x << "/" << pathLength << " t: " << t);
            //CI_LOG_D("added point x: " << smthPoint << " thick: " << thick);
        }
        
        mGeometryChanged = true;
        mInputPoints.push_back(newPos);
        mBounds.include(newPos);
    }
}
    
void Stroke::end()
{
    if (!mDrawing) return;
    
    mDrawing = false;
    mEndTime = app::getElapsedSeconds();
    
    updateGeometry();
}
    
void Stroke::update()
{
    if (!mDrawing && !mOptionsChanged && !mGeometryChanged) return;
    
    mOptionsChanged = false;
    mGeometryChanged = false;
    
    updateGeometry();
}
    
    
void Stroke::updateGeometry()
{
    if (!mShader) {
        mShader = gl::getStockShader(gl::ShaderDef().color());
    }

    int numPoints = mSampledPoints.size();
   
    if (numPoints > 0) {
        mNumUsedVertices = 0;
        
        // create necessary meshes
        // account for one point overlap between meshes
        int extraPoints = numPoints / mPointsPerMesh;
        int meshesNeeded = 1 + (extraPoints + numPoints) / mPointsPerMesh;
        
        while (mVboMeshes.size() < meshesNeeded ) {
            createMesh();
        }
        
        // update vertex positions
        // (we should only need to do this for the last mesh)
        for (int i = 0; i < mVboMeshes.size(); i++) {
        
            int start = i * (mPointsPerMesh - 1);
            int end = glm::min(start + mPointsPerMesh, (int) mSampledPoints.size());
            
            auto vertPosIter = mVboMeshes.at(i)->mapAttrib2f(geom::POSITION);
            
            for (int j = start; j < end; j++) {
                float p = glm::min(j, numPoints - j - 1);
                float x = 1.0f - glm::clamp(p / float(mOpts.lineCapPoints), 0.0f, 1.0f);
                
                float lineCap = glm::sqrt(1 - pow(x, 2));
                float thick = clamp(mOpts.baseThickness + mThickness.at(j) * mOpts.speedToThickness, mOpts.minThickness, mOpts.maxThickness);
                
                *vertPosIter = mSampledPoints.at(j) - mOffsets.at(j) * lineCap * thick;
                ++vertPosIter;
                
                *vertPosIter = mSampledPoints.at(j) + mOffsets.at(j) * lineCap * thick;
                ++vertPosIter;
                
                mNumUsedVertices += 2;
            }
            
            vertPosIter.unmap();
        }
    }
}
    
void Stroke::createMesh()
{
    // create a new mesh with all vertices collapsed
    std::vector<vec2> vertices(mVerticesPerMesh);
    
    auto layout = gl::VboMesh::Layout().usage(GL_DYNAMIC_DRAW).attrib(geom::Attrib::POSITION, 2);
    
    auto mesh = gl::VboMesh::create( mVerticesPerMesh, GL_TRIANGLE_STRIP, { layout } );
    mesh->bufferAttrib( geom::POSITION, sizeof(vec2) * vertices.size(), vertices.data() );
    
    mVboMeshes.push_back(mesh);
    mBatches.push_back(gl::Batch::create(mesh, mShader));
    
    mDebugColors.emplace_back(ColorA(CM_HSV, randFloat(), 1.0, 1.0, 0.5));
}
    
void Stroke::draw()
{
    int totalVertices = float(mNumUsedVertices) * mEndPct();
    
    for (int i = 0; i < mBatches.size(); i++) {
        int count = glm::min(mVerticesPerMesh , glm::max(totalVertices - i * mVerticesPerMesh, 0));
        int start = glm::max(int(float(totalVertices) * (mStartPct())) - i * mVerticesPerMesh, 0);
        mBatches[i]->draw(start, count - start);
    }

	//probably do this as a batch instead
	float maxPoints = 10.0f;
	if (mInputPoints.size() < maxPoints) {
		float size = 2; // *((maxPoints / mInputPoints.size()) / maxPoints);
		gl::drawSolidEllipse(mInputPoints.back(), size, size, 10);
	}
}

void Stroke::drawDebug()
{
    // draw bounding box
    gl::ScopedColor gray(Color::gray(0.75));
    gl::drawStrokedRect(mBounds);
    
    // draw mesh wireframe
    gl::enableWireframe();
    int totalVertices = float(mNumUsedVertices) * mEndPct();
    
    for (int i = 0; i < mBatches.size(); i++) {
        int count = glm::min(mVerticesPerMesh , glm::max(totalVertices - i * mVerticesPerMesh, 0));
        int start = glm::max(int(float(totalVertices) * (mStartPct())) - i * mVerticesPerMesh, 0);
        // int start = glm::max(int(float(totalVertices) * (mStartPct())) - i * count, 0);
        
        gl::ScopedColor col(mDebugColors[i]);
        mBatches[i]->draw(start, count - start);
    }
    gl::disableWireframe();
    
    // draw sampled points
    auto points = gl::VertBatch( GL_POINTS );
    for (auto &pos : mSampledPoints) {
        points.vertex(pos);
    }
    
    gl::pointSize(3.0);
    gl::ScopedColor white(1, 1, 1);
    points.draw();
    gl::pointSize(1.0);
}
    
void Stroke::animateIn(float duration, float delay)
{
    if (mDrawing) return;
    
    mAnimating = true;
    mStartPct.stop();
    mEndPct.stop();
    mStartPct = 0.0f;
    mEndPct = 0.0f;
    
    app::timeline().apply(&mEndPct, 1.0f, duration).delay(delay).finishFn([this] { mAnimating = false; });
}

void Stroke::animateOut(float duration, float delay)
{
    if (mDrawing) return;
    
    mAnimating = true;
    mStartPct.stop();
    mEndPct.stop();
    mStartPct = 0.0f;
    mEndPct = 1.0f;
    
    app::timeline().apply(&mStartPct, 1.0f, duration).delay(delay).finishFn([this] { mAnimating = false; });
}

} // namespace rph

File: Strokes/Assets/strokes/Stroke.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;



[System.Serializable]
public class Stroke {
	public Mesh mesh;
    int mVerticesPerMesh = 600;
    int mPointsPerMesh = 600 / 2;
    float mMinSegmentLength = .1f;//.0f;

    bool mDrawing = false;
    bool mOptionsChanged = false;
    bool mGeometryChanged = false;

    int mId = -1;
    int mNumUsedVertices = 0;
    float mStrokeLength = 0;

    Rect mBounds;
    public float baseThickness = .01f;
	public float minThickness = .01f;
	public float maxThickness = 2f;
	public float speedToThickness = 2f;
	public int lineCapPoints = 10;
    List<Vector3> mInputPoints;
    List<Vector3> mSampledPoints;
    List<Vector3> mOffsets;
    List<float> mThickness;
    
	public Stroke(Mesh _mesh){
        //initialize all lists
        mInputPoints = new List<Vector3>();
        mSampledPoints= new List<Vector3>();
        mOffsets= new List<Vector3>();
        mThickness= new List<float>();
        mesh = _mesh;
	}
    public void BuildStroke(List<Vector3> _points){
        foreach (Vector3 point in _points)
        {
            if (mInputPoints.Count == 0)
            {
                start(point);
            }
            else
            {
                move(point);
            }
        }
        end();
	}
public void start(Vector3 pos)
{
    mDrawing = true;
    move(pos);
}

public void move(Vector3 pos)
{
    if (!mDrawing) return;

    if (mInputPoints.Count <= 2)
    {
        mInputPoints.Add(pos);
    }
    else
    {
		Vector3 lastPos =mInputPoints[mInputPoints.Count - 1];
        Vector3 newPos = Vector3.Lerp(lastPos, pos, .35f);

        // calculate thickness
        float dist = Vector3.Distance(newPos, lastPos);

        if (dist < .1) return;

    //     //        float newThick = clamp(mOpts.baseThickness + dist * mOpts.speedToThickness, mOpts.minThickness, mOpts.maxThickness);
        float newThick = dist;
        float prevThick = mThickness.Count <1 ? newThick : mThickness[mThickness.Count -1];
        float currThick = Mathf.Lerp(prevThick, newThick, 0.2f); //lerp(prevThick, newThick, 0.2);

        // get last 3 input points
		Vector3 prev2 = mInputPoints[mInputPoints.Count-2];
		Vector3 prev1 = mInputPoints[mInputPoints.Count - 1];
		Vector3 cur = newPos;

		//come back to this bit
		//create bezier segment with inputs as control points
		Vector3[] pathPnts =new Vector3[3];
		pathPnts[0] = (prev2 + prev1) / 2.0f;
		pathPnts[1] = prev1;
		pathPnts[2] = (prev1 + cur) / 2.0f;
		CubicBezierPath path = new CubicBezierPath(pathPnts);

        // divide segment adaptatively depending on its length
        // save vertices and thickness
        float pathLength = path.ComputeApproxLength();
        int divisions = (int)(pathLength / mMinSegmentLength);

        for (int i = 1; i <= divisions; i++)
        {
            float t = i / (float)divisions;
            float thick = Mathf.Lerp(prevThick, currThick, t);
            Vector3 point = path.GetPointNorm(t);
			Vector3 norm = Vector3.Normalize( newPos - (mSampledPoints.Count>0 ? mSampledPoints[mSampledPoints.Count-1] : mInputPoints[mInputPoints.Count-1]));
			Vector3 perp = new Vector3(-norm.y, norm.x, norm.z);

            mSampledPoints.Add(point);
            mThickness.Add(thick);
            mOffsets.Add(perp);
        }

        mGeometryChanged = true;
        mInputPoints.Add(newPos);
		//update the bounds
        updateMesh();
    }
}

public void end()
{
    if (!mDrawing) return;

    mDrawing = false;
	Debug.Log("end");
    updateMesh();
}
	void updateMesh(){

        int numPoints = mSampledPoints.Count;

        if (numPoints > 0)
        {
            mNumUsedVertices = 0;
			Debug.Log("mesh numPoints:"+numPoints);

            //build the vertices
            Vector3[] vertices = new Vector3[numPoints*2];
            for (int i = 0; i < numPoints; i++)
            {
                
                float p = Mathf.Min(i, numPoints - i - 1);
                float x = 1.0f -Mathf.Clamp(p / (float)lineCapPoints, 0.0f, 1.0f);
                float lineCap = Mathf.Sqrt(1 - Mathf.Pow(x, 2));
                // float lineCap = 1;

                float thick = Mathf.Clamp(baseThickness + mThickness[i] * speedToThickness, minThickness, maxThickness);
                //set the vertices:
                //base point - the perpendicular offset * thickness
                vertices[i*2] = mSampledPoints[i]-mOffsets[i] *thick*lineCap;
                //base point + the perpendicular offset *thickness
                vertices[i * 2 +1] = mSampledPoints[i] + mOffsets[i] * thick*lineCap;
            }
            mesh.vertices = vertices;

            //set uvs
            Vector2[] uvs = new Vector2[vertices.Length];
            for (int k = 0; k < uvs.Length; k+=2)
            {
                uvs[k] = new Vector2(k/(float)uvs.Length, 0);
                uvs[k+1] = new Vector2(k/(float)uvs.Length, 1);
            }
            mesh.uv = uvs;

            //build the triangles
            int[] triangles = new int[(numPoints * 2 - 2) * 3];
            int j = 0;
            for (int i = 0; i < numPoints * 2 - 3; i += 2, j++)
            {
                triangles[i * 3] = j * 2;
                triangles[i * 3 + 1] = j * 2 + 1;
                triangles[i * 3 + 2] = j * 2 + 2;

                triangles[i * 3 + 3] = j * 2 + 1;
                triangles[i * 3 + 4] = j * 2 + 3;
                triangles[i * 3 + 5] = j * 2 + 2;

            }
            mesh.triangles = triangles;

            mesh.RecalculateNormals();
        }
	}
}
